src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.cpp

```diff
@@ -61,10 +65,119 @@ double overlap_width(const size_t overlap_extent,
   return collocation_points[overlap_extent] - collocation_points[0];
 }
 
+template <size_t Dim>
+OverlapIterator::OverlapIterator(const Index<Dim>& volume_extents,
+                                 const size_t overlap_extent,
+                                 const Direction<Dim>& direction) noexcept
+    : size_{overlap_num_points(volume_extents, overlap_extent,
+                               direction.dimension())},
+      num_slices_{overlap_extent},
+      stride_{std::accumulate(volume_extents.begin(),
+                              volume_extents.begin() + direction.dimension(),
+                              1_st, std::multiplies<size_t>())},
+      stride_count_{0},
+      jump_{(volume_extents[direction.dimension()] - num_slices_) * stride_},
+      initial_offset_{stride_ * (direction.side() == Side::Lower
+                                     ? 0
+                                     : (volume_extents[direction.dimension()] -
+                                        num_slices_))},
+      volume_offset_{initial_offset_},
+      overlap_offset_{0} {
+  assert_overlap_extent(volume_extents, overlap_extent, direction.dimension());
+}
+
+OverlapIterator::operator bool() const noexcept {
+  return overlap_offset_ < size_;
+}
+
+OverlapIterator& OverlapIterator::operator++() {
+  ++volume_offset_;
+  ++overlap_offset_;
+  ++stride_count_;
+  if (stride_count_ == stride_ * num_slices_) {
+    volume_offset_ += jump_;
+    stride_count_ = 0;
+  }
+  return *this;
+}
+
+size_t OverlapIterator::volume_offset() const noexcept {
+  return volume_offset_;
+}
+
+size_t OverlapIterator::overlap_offset() const noexcept {
+  return overlap_offset_;
+}
+
+void OverlapIterator::reset() noexcept {
+  volume_offset_ = initial_offset_;
+  overlap_offset_ = 0;
+  stride_count_ = 0;
+}
+
+namespace detail {
+
+template <size_t Dim>
+void data_on_overlap_impl(double* overlap_data, const double* volume_data,
+                          const size_t num_components,
+                          const Index<Dim>& volume_extents,
+                          const size_t overlap_extent,
+                          const Direction<Dim>& direction) noexcept {
+  const size_t volume_num_points = volume_extents.product();
+  const size_t overlap_num_points = LinearSolver::Schwarz::overlap_num_points(
+      volume_extents, overlap_extent, direction.dimension());
+  for (OverlapIterator overlap_iterator{volume_extents, overlap_extent,
+                                        direction};
+       overlap_iterator; ++overlap_iterator) {
+    for (size_t j = 0; j < num_components; ++j) {
+      const size_t volume_data_index =
```

Comment:
Not a change request, just a note in case you find yourself needing to write this type of code often -- If `OverlapIterator` was an STL-esque iterator associated with a particular container, and with `operator*` accessing elements from the volume, these helper functions could just be `std::copy` and `std::transform` calls.

Reply:
If I understand things correctly, that would require an extra wrapper class around the volume data that implements the `operator*`. It's an intriguing idea to reduce the implementation to `std::copy` though, I'll keep this in mind.

Reply:
I agree it would involve some restructuring, which I didn't want to request for this PR. In principle, you could also pass into an overloaded constructor of this iterator the volume data it was to index into, or have a 'set_data' function that does the same, then create the `operator*` for that.